src/memory/mmap_bytebuf.rs

```diff
@@ -5,181 +5,125 @@
-        }
+        let allocation =
+            region::alloc(len, Protection::READ_WRITE).map_err(|error| error.to_string())?;
+        Ok(Self { allocation })
     }
 
-    fn as_slice_mut(&mut self) -> &mut [u8] {
-        unsafe {
-            // Safety Proof:
-            // - See the proof for `Self::as_slice`
-            // - Additionally, it is not possible to obtain two mutable references for `self.ptr`
-            slice::from_raw_parts_mut(self.ptr.as_ptr(), self.len)
-        }
+    /// Returns a shared slice over the bytes of the virtual memory allocation.
+    #[inline]
+    pub fn as_slice(&self) -> &[u8] {
+        // # SAFETY
+        //
+        // The operation is safe since we assume that the virtual memory allocation
+        // has been successful and allocated exactly `self.allocation.len()` bytes.
+        // Therefore creating a slice with `self.len` elements is valid.
+        // Aliasing guarantees are not violated since `self` is the only owner
+        // of the underlying virtual memory allocation.
+        unsafe { slice::from_raw_parts(self.allocation.as_ptr(), self.allocation.len()) }
     }
-}
 
-impl Drop for Mmap {
-    fn drop(&mut self) {
-        let ret_val = unsafe {
-            // Safety proof:
-            // - `self.ptr` was allocated by a call to `mmap`.
-            // - `self.len` was saved at the same time and it doesn't change throughout the lifetime
-            //   of `self`.
-            libc::munmap(self.ptr.as_ptr() as *mut libc::c_void, self.len)
-        };
-
-        // There is no reason for `munmap` to fail to deallocate a private annonymous mapping
-        // allocated by `mmap`.
-        // However, for the cases when it actually fails prefer to fail, in order to not leak
-        // and exhaust the virtual memory.
-        assert_eq!(ret_val, 0, "munmap failed");
+    /// Returns an exclusive slice over the bytes of the virtual memory allocation.
+    #[inline]
+    pub fn as_slice_mut(&mut self) -> &mut [u8] {
+        // # SAFETY
+        //
+        // See safety proof of `Mmap::as_slice`.
+        // Additionally, it is not possible to obtain two mutable references for the same memory area.
+        unsafe { slice::from_raw_parts_mut(self.allocation.as_mut_ptr(), self.allocation.len()) }
     }
 }
 
+/// A virtually allocated byte buffer.
 pub struct ByteBuf {
-    mmap: Option<Mmap>,
+    /// The underlying virtual memory allocation.
+    mem: VirtualMemory,
+    /// The current size of the used parts of the virtual memory allocation.
+    len: usize,
 }
 
 impl ByteBuf {
+    /// Determines the initial size of the virtual memory allocation.
+    ///
+    /// # Note
+    ///
+    /// In this implementation we won't reallocate the virtually allocated
+    /// buffer and instead simply adjust the `len` field of the `ByteBuf`
+    /// wrapper in order to efficiently grow the virtual memory.
+    const ALLOCATION_SIZE: usize = u32::MAX as usize;
```

Comment:
what would you recommend then? 2GB? or a whole different approach? note that most platforms where `wasmi` runs on are 64-bit so we should keep our focus on those with respect to performance.

Reply:
While I agree with you that 64-bit is our primary target, I think the compatibility is also important. I think a marginal win in performance does not outweigh the feature of running on 32-bit environments.

There are a couple of possible ways forward:

- Either return to the previous approach where we allocate as needed, or
- fallback in the case of 32-bit environments. We could, for example, fallback on the allocate as needed for them, or just plainly fallback to the vec impl.

Reply:
Reallocating the virtual memory + copying over all its contents seems to be destroying the purpose of using virtual memory in the first place so I'd simply fallback on 32-bit platforms for now.

Reply:
I agree. We should just fall back to the `Vec` implementation on 32-bit platforms. Otherwise we would need to provide a configuration option for how much of the virtual memory space we are allowed to take per instance.

Reply:
done